combine filter: fine list handling (option b) (#57894)
ansible/ansible: docs/docsite/rst/user_guide/playbooks_filters.rst

.. versionadded:: 2.0

The `combine` filter allows hashes to be merged.
For example, the following would override keys in one hash::

    {{ {'a':1, 'b':2} | combine({'b':3}) }}

The resulting hash would be::

    {'a':1, 'b':3}

The filter can also take multiple arguments to merge::

    {{ a | combine(b, c, d) }}
    {{ [a, b, c, d] | combine }}

> In this case, keys in `d` would override those in `c`, which would

Reviewer: ```suggestion
In this case, keys in ``d`` would override those in ``c``, which would
```
This is Restructured Text, so use double backticks for code formatting.

Author: done (plus all the other ` in the combine section)

override those in ``b``, and so on.

The filter also accepts two optional parameters: ``recursive`` and ``list_merge``.

recursive
  Is a boolean, default to ``False``.
  Should the ``combine`` recursively merge nested hashes.
  Note: It does **not** depend on the value of the ``hash_behaviour`` setting in ``ansible.cfg``.

list_merge

>   Is a string, it's possible values are ``replace`` (default), ``keep``, ``append``, ``prepend``, ``append_rp`` or ``prepend_rp``.

Reviewer: ```suggestion
  Is a string, its possible values are ``replace`` (default), ``keep``, ``append``, ``prepend``, ``append_rp`` or ``prepend_rp``.
```

Author: done

>   It modify the behaviour of ``combine`` when the hashes to merge contains arrays/lists.

Reviewer: ```suggestion
  It modifies the behaviour of ``combine`` when the hashes to merge contain arrays/lists.
```

Author: done

    {{ default | combine(patch, recursive=True) }}

This would result in::

    a:
      x: default
      y: patch
      z: patch
    b: patch
    c: default

If ``list_merge='replace'`` (the default), arrays from the right hash will "replace" the ones in the left hash::

    default:
      a:
        - default
    patch:
      a:
        - patch

.. code-block:: jinja

    {{ default | combine(patch) }}

This would result in::

    a:
      - patch

If ``list_merge='keep'``, arrays from the left hash will be kept::

    {{ default | combine(patch, list_merge='keep') }}

This would result in::

    a:
      - default

If ``list_merge='append'``, arrays from the right hash will be appended to the ones in the left hash::

    {{ default | combine(patch, list_merge='append') }}

This would result in::

    a:
      - default
      - patch

If ``list_merge='prepend'``, arrays from the right hash will be prepended to the ones in the left hash::

    {{ default | combine(patch, list_merge='prepend') }}

This would result in::

    a:
      - patch
      - default

If ``list_merge='append_rp'``, arrays from the right hash will be appended to the ones in the left hash.

> And elements of the left hash's array that are also in the right hash's array will be remove ("rp" stands for "remove present").

Reviewer: ```suggestion
Elements of arrays in the left hash that are also in the corresponding array of the right hash will be removed ("rp" stands for "remove present").
```

Author: done

Duplicate elements that aren't in both hashes are kept::

    default:
      a:
        - 1
        - 1
        - 2
        - 3
    patch:
      a:
        - 3
        - 4
        - 5
        - 5

.. code-block:: jinja

    {{ default | combine(patch, list_merge='append_rp') }}

This would result in::

    a:
      - 1
      - 1
      - 2
      - 3
      - 4
      - 5
      - 5

> If ``list_merge='prepend_rp'``, is the same as ``append_rp`` but elements of the right hash's array are prepend::

Reviewer: ```suggestion
If ``list_merge='prepend_rp'``, the behavior is similar to the one for ``append_rp``, but elements of arrays in the right hash are prepended::
```

Author: done
